java/src/main/java/com/google/oak/transport/GrpcStreamingTransport.java

```diff
@@ -0,0 +1,138 @@
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.function.Function;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import oak.session.noninteractive.v1.AttestationBundle;
+import oak.session.noninteractive.v1.GetPublicKeyRequest;
+import oak.session.noninteractive.v1.GetPublicKeyResponse;
+import oak.session.noninteractive.v1.InvokeRequest;
+import oak.session.noninteractive.v1.InvokeResponse;
+import oak.session.noninteractive.v1.RequestWrapper;
+import oak.session.noninteractive.v1.ResponseWrapper;
+
+public class GrpcStreamingTransport implements EvidenceProvider, Transport {
+  private static final Logger logger = Logger.getLogger(GrpcStreamingTransport.class.getName());
+
+  private BlockingQueue<ResponseWrapper> messageQueue;
+  private StreamObserver<RequestWrapper> requestObserver;
+
+  /**
+   * Creates an instance of {@code GrpcStreamingTransport}.
+   *
+   * @param stream a method reference to a gRPC client streaming method with the appropriate request
+   * and response types.
+   */
+  public GrpcStreamingTransport(
+      Function<StreamObserver<ResponseWrapper>, StreamObserver<RequestWrapper>> stream) {
+    // We expect to receive a single response message for each request.
+    this.messageQueue = new ArrayBlockingQueue<>(1);
+    StreamObserver<ResponseWrapper> responseObserver = new StreamObserver<ResponseWrapper>() {
+      @Override
+      public void onNext(ResponseWrapper response) {
+        try {
+          messageQueue.put(response);
+        } catch (Exception e) {
+          if (e instanceof InterruptedException) {
+            Thread.currentThread().interrupt();
+          }
+          logger.log(Level.WARNING, "Couldn't send server response to the message queue: " + e);
+        }
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        Status status = Status.fromThrowable(t);
+        logger.log(Level.WARNING, "Couldn't receive response: " + status);
+      }
+
+      @Override
+      public void onCompleted() {
+        logger.log(Level.FINE, "response message queue completed");
+      }
+    };
+    this.requestObserver = stream.apply(responseObserver);
+  }
+
+  /**
+   * Returns evidence about the trustworthiness of a remote server.
+   *
+   * @return {@code AttestationBundle} wrapped in a {@code Result}
+   */
+  public Result<AttestationBundle, Exception> getEvidence() {
+    try {
+      RequestWrapper requestWrapper = RequestWrapper.newBuilder()
+                                          .setGetPublicKeyRequest(GetPublicKeyRequest.newBuilder())
+                                          .build();
+      logger.log(Level.INFO, "sending get evidence request: " + requestWrapper);
+      this.requestObserver.onNext(requestWrapper);
+
+      ResponseWrapper responseWrapper = this.messageQueue.poll(10, SECONDS);
```

Comment:
Where does 10 seconds come from? This feels arbitrary. Should not polling be retried?

Reply:
Added a TODO for adding retry